File: contest/b_the_parrots_of_tortuga/sol/sol.cs
using System;

public class b_the_parrots_of_tortuga
{
    public static void Main()
    {
        int nb_parrots;
        int best_offer = 0;
        String seller = "";
        
        String[] parts = Console.ReadLine().Split();
        int g = Convert.ToInt32(parts[0]);
        int n = Convert.ToInt32(parts[1]);
        
        // Go through all trainers in the given order
        for(int i = 0; i < n; ++i){
            parts = Console.ReadLine().Split();
            String name = parts[0];
            int p = Convert.ToInt32(parts[1]);
            int c = Convert.ToInt32(parts[2]);
            
            // Compute how many parrots we can buy from this trainer
            nb_parrots = (int)Math.Floor((double)((g - c) / p));

            // Check if it's better than the best encountered yet
            if( nb_parrots > best_offer ){
                best_offer = nb_parrots;
                seller = name;
            }
        }
        
        if(best_offer == 0){
            Console.WriteLine("Impossible");
        }else{
            Console.WriteLine(best_offer);
            Console.WriteLine(seller);
        }
        
    }
}

File: contest/b_the_parrots_of_tortuga/sol/sol.php
<?php

$best_offer = 0;
$seller = "";

$line = trim(fgets(STDIN));
$parts = explode(" ", $line);
$g = intval($parts[0]);
$n = intval($parts[1]);

// Go through all trainers in the given order
for($i = 0; $i < $n; $i++){
    $line = trim(fgets(STDIN));
    $parts = explode(" ", $line);
    $name = $parts[0];
    $p = intval($parts[1]);
    $c = intval($parts[2]);
    
    // Compute how many parrots we can buy from this trainer
    $nb_parrots = floor(($g - $c) / $p);

    // Check if it's better than the best encountered yet
    if( $nb_parrots > $best_offer ){
        $best_offer = $nb_parrots;
        $seller = $name;
    }
}

if($best_offer>0){
    echo $best_offer . "\n";
    echo $seller . "\n";
}else{
    echo "Impossible" . "\n";
}
